### PR TITLE
StateTimeline: Show more precise duration

### LINKED ISSUE #42269: "State Timeline" durations on tooltips not informative (floor rounding)
**Steps to reproduce**:

1. Add "State Timeline" chart 
2. Enable tooltips
3. Add data with duration more 1 hour and close to 2. I.e. `1 h 10 m` and `1 h 55 m`
4. Hover intervals

**Actual Result**:

Before hour, it shows values convenient like `30 minutes`, `40 minutes`.  
Just unusable value `1 hour` for intervals more than hour. Seems floor round.
So you cannot fast analyze data. I.e. (see screencast) almost all gaps shows as `1 hour`,
Despite values can be even `1h40`. So it hard to easily found disperse by quick hover.
And data on chars is perceived quite difficult.

**Expected Result**

* Show more precise values like `1 hour 50 min` 
* (Just as idea) Maybe add some setting for `State Timeline` charts or somehow in `[date_formats]`([like this](https://grafana.com/docs/grafana/latest/administration/configuration/#date_formats)). But not sure how it should look and work. Just intervals are dynamic and can be around `days`, `seconds` or even `months`. 
  

**Relevant WCAG Criteria:** [#.#.# WCAG Criterion](link to https://www.w3.org/WAI/WCAG21/quickref/?versions=2.0)
n/a

**Environment**:
- Grafana version: v8.3.0-beta1 (63cb359220) 
- Data source type & version: Prometheus
- OS Grafana is installed on: (docker version: grafana/grafana:8.3.0-beta1)
- User OS & Browser: Ubuntu 20.04.3 & Brave version 1.31.91 (Chromium: 95.0.4638.69)
- Grafana plugins: 
- Others

image

StateTimeline_durations

### PR DESCRIPTION
**Pull request by zoltanbedi**
**What this PR does / why we need it**:

The state timeline tooltip shows the duration to only 1 precision point which means that it can be confusing to see that the duration is 1 hour when it is 1 hour and 50 minutes. I've added 1 more precision point so it will show minutes as well. 

Screenshot 2022-01-14 at 14 44 13

**Which issue(s) this PR fixes**:

Fixes #42269 

**Special notes for your reviewer**:

Should we make this customisable?

**To reproduce**:
<details>
<summary>Dashboard JSON</summary>

```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 2,
  "id": 205,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "fillOpacity": 70,
            "lineWidth": 0
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "#EAB839",
                "value": 0
              }
            ]
          },
          "unit": "string"
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "alignValue": "left",
        "legend": {
          "displayMode": "list",
          "placement": "bottom"
        },
        "mergeValues": false,
        "rowHeight": 0.9,
        "showValue": "always",
        "tooltip": {
          "mode": "single",
          "sort": "none"
        }
      },
      "targets": [
        {
          "csvContent": "time,value\n1642142562605, LOW\n1642147962605, HIGH\n1642154362605, LOW",
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "refId": "A",
          "scenarioId": "csv_content"
        }
      ],
      "title": "Duration",
      "type": "state-timeline"
    }
  ],
  "refresh": "",
  "schemaVersion": 34,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-12h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "state timeline",
  "uid": "HD3XieDnk",
  "version": 7,
  "weekStart": ""
}
```
</details>

### CODE REVIEW

**Comment by dprokop** [comment]
Don't have thoughts yet, but initial feedback is that this probably ain't enough. See the data below (states spanning over months), we are still missing the precision:

```csv
time,value
1609459200000, LOW
1617235200000, HIGH
1625491500000, LOW
1630501500000,HIGH
````

**Comment by ryantxu** [comment]
> Maybe the precision should be controlled via field config?

Lets avoid if possible :)

> Or alternatively we could be slightly more clever and automate the precision based on the minmal step of the time field.

This reminds me of a LONG ago requested feature that never got merged :cry: 
https://github.com/grafana/grafana/pull/11878

image

**Comment by leeoniya** [comment]
certainly feels like there should be some kind of equivalent to `.toPrecision(3)` for time durations.

we should try to do show something like 3 elements of `years, months, weeks, days, hours, minutes, secs, ms` based on the range. maybe Moment has something like this built in already, or https://github.com/hustcc/timeago.js, or we can just hand-roll a small implementation, similar to 

https://github.com/grafana/grafana/blob/7422789ec7ab665fd5d01e3b891f44ac427061c5/public/app/features/explore/Time.tsx#L14

or

https://github.com/grafana/grafana/blob/7422789ec7ab665fd5d01e3b891f44ac427061c5/packages/grafana-ui/src/components/uPlot/config/UPlotAxisBuilder.ts#L188

**Comment by leeoniya** [comment]
here's my version (with years, months, weeks support):

```js
// adapted from https://gist.github.com/remino/1563878
function fmtDuration(milli) {
  if (milli < 0 || Number.isNaN(milli)) {
    return null;
  }

  let yr, mo, wk, d, h, m, s, ms;

  s = Math.floor(milli / 1000);
  m = Math.floor(s / 60);
  s = s % 60;
  h = Math.floor(m / 60);
  m = m % 60;
  d = Math.floor(h / 24);
  h = h % 24;

  yr = Math.floor(d / 365);
  if (yr > 0)
    d = d % 365;

  mo = Math.floor(d / 30.5);
  if (mo > 0)
    d = d % 30;

  wk = Math.floor(d / 7);
  if (wk > 0)
    d = d % 7;

  ms = Math.floor((milli % 1000) * 1000) / 1000;

  return (
    yr > 0 ? yr + 'y '  + (mo > 0 ? mo + 'mo ' : '') + (wk > 0 ? wk + 'w ' : '') + (d > 0 ? d  + 'd ' : '') :
    mo > 0 ? mo + 'mo ' + (wk > 0 ? wk + 'w '  : '') + (d  > 0 ? d  + 'd ' : '') :
    wk > 0 ? wk + 'w '  + (d  > 0 ? d  + 'd '  : '') :
    d  > 0 ? d  + 'd '  + (h  > 0 ? h  + 'h '  : '') :
    h  > 0 ? h  + 'h '  + (m  > 0 ? m  + 'm '  : '') :
    m  > 0 ? m  + 'm '  + (s  > 0 ? s  + 's '  : '') :
    s  > 0 ? s  + 's '  + (ms > 0 ? ms + 'ms ' : '') :
    ms > 0 ? ms + 'ms ' :
    '0'
  ).trim();
}
```

**Comment by zoltanbedi** [comment]
@leeoniya I added your code to the state timeline's util file and changed it slightly for months to calculate with 28 days instead of 30.5. Let me know if you think that doesn't make sense.

File: public/app/plugins/panel/state-timeline/utils.ts [hunk 1]
```diff
@@ -567,3 +567,66 @@ export function findNextStateIndex(field: Field, datapointIdx: number) {
...
+  }
+
+  wk = Math.floor(d / 7);
+
+  if (wk > 0) {
+    d = d % 7;
+  }
+
+  ms = Math.round((milliSeconds % 1000) * 1000) / 1000;
+  console.log({ ms, s, m, h, d, wk, mo, yr });
```
Comment (leeoniya) [inline public/app/plugins/panel/state-timeline/utils.ts:612]: oops

File: public/app/plugins/panel/state-timeline/utils.ts [hunk 2]
```diff
@@ -567,3 +567,66 @@ export function findNextStateIndex(field: Field, datapointIdx: number) {
...
+  m = m % 60;
+  d = Math.floor(h / 24);
+  h = h % 24;
+
+  yr = Math.floor(d / 365);
+  if (yr > 0) {
+    d = d % 365;
+  }
+
+  console.log({ d });
```
Comment (leeoniya) [inline public/app/plugins/panel/state-timeline/utils.ts:598]: oops

**Comment by leeoniya** [comment]
> calculate with 28 days instead of 30.5. Let me know if you think that doesn't make sense.

11 * 28 is quite short of 11 months. i think 30 or 30.5 makes more sense here. 🤷 

also might be worth adding a couple tests that skip a segment. e.g. "1y 1w" instead of what we had before like "1y 0mo 1w". this code was made to omit any 0 intermediates.

do you think this util fn should be moved out to some general date utils or keep local to this vis?

File: public/app/plugins/panel/state-timeline/utils.ts (hunk 2 above)
Comment (zoltanbedi) [reply public/app/plugins/panel/state-timeline/utils.ts:598]: oops 😄

**Comment by zoltanbedi** [comment]
> > calculate with 28 days instead of 30.5. Let me know if you think that doesn't make sense.
> 
> 11 * 28 is quite short of 11 months. i think 30 or 30.5 makes more sense here. 🤷
> 
> also might be worth adding a couple tests that skip a segment. e.g. "1y 1w" instead of what we had before like "1y 0mo 1w". this code was made to omit any 0 intermediates.
> 
> do you think this util fn should be moved out to some general date utils or keep local to this vis?

I choose 28 days as it was kind of weird for me that for 28 days it says that 4 weeks instead of 1 month but 30 days makes sense for me as well.

Added some more tests.

So I was thinking about this and the reason I choose to put it next to the vis is because we already have a toDuration function with parameters. People would be confused which one to use. But I'm up for moving this out from here.

**Comment by leeoniya** [comment]
> So I was thinking about this and the reason I choose to put it next to the vis is because we already have a toDuration function with parameters. People would be confused which one to use. But I'm up for moving this out from here.

sounds good. toDuration() could use a refactor and maybe we can merge these in the future.

**Review by dprokop — APPROVED** [review]
Now it works way better @zoltanbedi!

**Pull request merged**